### PR TITLE
feat: impl contract calls

### PR DESCRIPTION
**Pull request by 0xrusowsky**
this PR brings x-contract calls to r55.

## interface macro
the most opinionated design choice that i made is to create a macro (inspired by the contract macro) that based on an interface (trait) generates the necessary syntactic sugar to call a contract, handling abi en/decoding and the risc-v runtime call.

as u can see in `erc20x` the user simply defines an interface as they would do in solidity
https://github.com/0xrusowsky/r55/blob/110cc5addd481368723eb0b992f081231c2b8d7d/erc20x/src/lib.rs#L17-L20

and then they can use it by initializing with the address of the target contract:
https://github.com/0xrusowsky/r55/blob/110cc5addd481368723eb0b992f081231c2b8d7d/erc20x/src/lib.rs#L25-L29

i wanted to leverage rust's type system, so the macro forces all methods of the interface to return options (in the future it would be cool to support custom errors). As u can see in the example above, unlike with solidity, we force devs to handle potential call errors.

## r55 execution
after i had the interface implementation, i had to integrate the calls with revm.
in this case i've had to touch code that was written by others beforehand, so i'll try to justify why i changed some of the stuff that had been previously done >> the main reason is that frames (how revm handles the execution context of each call) hadn't been fully taken into account.

relevant changes that i made:
- r55 doesn't handle "evm gas" anymore, as the cumulated amounts were lost between calls. Instead, the gas computation is delegated to the revm interpreter >> i implemented the `syscall_gas!` macro (maybe not the best name), which allows us to simply worry about the riscv gas. Finally, since i was touching gas, i also moved away from magic numbers and started using constants.
-  r55 totally ignores revm shared memory (as it uses its on DRAM). The main change here was adding memory management in the handle register. In this case, when the interpreter action is a result (not the last one), we need to write the output to the memory return data destiny that the parent frame is expecting to read from.
- in the frame call stack, rather than always getting the first, we had to get the last one, so that it acts like a stack (lifo)

## logging and tests
in order to make it work, i had to debug a lot, so i ended up having prints everywhere lol
however, i thought that some of those where quite useful to keep, specially while there is still stuff to implement. Because of that i moved from `env_logger` to `tracing`, and added a bunch of `debug` logs (that won't run by default, but which are really useful when developing).

you can see the debug logs by running:

```
RUST_LOG=debug cargo test --package r55 --test e2e -- erc20 --exact
```

PS: i have format-on-save by default, so sorry for that, i know it's annoying.

### CODE REVIEW

File: r55/src/exec.rs [hunk 1]
```diff
@@ -133,32 +136,89 @@ pub fn handle_register<EXT, DB: Database>(handler: &mut EvmHandler<'_, EXT, DB>)
...
+                            debug!("- Memory size: {}", return_memory.len());
+
+                            // write return data to parent's memory
+                            if res.output.len() == return_memory.len() {
+                                return_memory.copy_from_slice(&res.output);
+                            } else {
+                                warn!("Unexpected return data size!");
+                            }
+                        }
+                    }
```
Comment (gakonst) [inline r55/src/exec.rs:188]: can u unpack a bit what's going on here

File: r55/src/exec.rs [hunk 2]
```diff
@@ -182,33 +241,23 @@ fn execute_riscv(
...
-                            return Ok(InterpreterAction::Return {
-                                result: InterpreterResult {
-                                    result: InstructionResult::OutOfGas,
-                                    output: Bytes::new(),
-                                    gas: interpreter.gas,
-                                },
-                            });
-                        }
+                        // RETURN logs the gas of the whole risc-v instruction set
+                        syscall_gas!(interpreter, r55_gas);
```
Comment (gakonst) [inline r55/src/exec.rs:258]: seems fine right?

Comment (gakonst) [inline r55/src/exec.rs:258]: seems fine right?

**Review by gakonst — COMMENTED** [review]
concept ack

File: r55/src/exec.rs (hunk 1 above)
Comment (0xrusowsky) [reply r55/src/exec.rs:188]: that's what i was trying to explain in the PR description regarding r55 memory.

my understanding is that in revm we have a concept of shared memory, whereas in this case we have DRAM serving the same purpose (hence why we don't need MLOAD and MSTORE).
since each frame (call) uses an independent riscv context with it's own emulator instance, we need to keep track of the expected return data range location in DRAM (`RVEmu.returned_data_destiny`).

when the parent [FRAME-0] executes a `CALL`, we initialize + cache the memory range:
https://github.com/0xrusowsky/r55/blob/7e8eebd29ce061aea541fe33707b0d00c7c0900f/r55/src/exec.rs#L350-#L356

the child [FRAME-1] computes the output of the call, but it is unaware of the expected return data destiny of the parent. Because of that i thought that the only place where i could "move data between frames" was in `fn handle_register()`, where we manipulate the call stack >> in the line where u added the comment is where i have access to both the returned data and the parent context at the same time.

posting a simplified version of the debug logs to help understand the explanation above
```
// emited in `handle_register` (call stack)
=== [FRAME-0] Contract: 0x114c28ea2CDe99e41D2566B1CfAA64a84f564caf ===========

// emited in `execute_riscv` (risc-v emu)
Starting RISC-V execution:  PC: 0x80300000  Return data dst: None
[Syscall::call - 0xf1]
> Return data will be written to: 2150651684..2150651716

// emited in `handle_register` (call stack)
=== [FRAME-1] Contract: 0xf6a171f57ACAc30C292e223eA8aDBb28abD3E14d ===========

// emited in `execute_riscv` (risc-v emu)
Starting RISC-V execution:  PC: 0x80300000  Return data dst: None
[Syscall::return - 0xf3]

// emited in `handle_register` (call stack) >>> this is the step where i copy the output data to the expected location by the parent frame
=== [FRAME-1] Ouput: RETURN
- Return data: 0x000000000000000000000000000000000000000000000000000000000000002a
- Memory range: 2150651684..2150651716
- Memory size: 32

// emited in `handle_register` (call stack)
=== [FRAME-0] Contract: 0x114c28ea2CDe99e41D2566B1CfAA64a84f564caf ===========

// emited in `execute_riscv` (risc-v emu)
Resuming RISC-V execution:  PC: 0x80300994  Return data dst: Some(
    2150651684..2150651716,
)
Loaded return data: 0x000000000000000000000000000000000000000000000000000000000000002

// emited in `handle_register` (call stack)
=== [FRAME-0] Ouput: RETURN (last)
```

File: r55/src/exec.rs (hunk 2 above)
Comment (0xrusowsky) [reply r55/src/exec.rs:258]: when testing i made sure that i got the exact same gas costs as the original implementation.
as mentioned above, this approach fixes the scenario where there are multiple calls (execution contexts)

File: r55/src/exec.rs (hunk 1 above)
Comment (0xrusowsky) [reply r55/src/exec.rs:188]: assuming that what i explained above makes sense... git blame was pointing to Dragan as the editor of most of the `handle_register()` code, so maybe u also want him to validate that my proposed approach is sound

File: r55/src/test_utils.rs [hunk 3]
```diff
@@ -8,7 +8,18 @@ static INIT: Once = Once::new();
...
     INIT.call_once(|| {
-        env_logger::builder().is_test(true).try_init().unwrap();
+        let log_level = std::env::var("RUST_LOG").unwrap_or("INFO".to_owned());
+        let tracing_sub = tracing_subscriber::fmt()
+            .with_max_level(tracing::Level::DEBUG)
+            .with_env_filter(tracing_subscriber::EnvFilter::new(&format!(
+                "{}",
+                log_level
+            )))
+            .with_target(false)
```
Comment (leonardoalt) [inline r55/src/test_utils.rs:18]: Why are these needed?

File: r55/src/test_utils.rs [hunk 4]
```diff
@@ -8,7 +8,18 @@ static INIT: Once = Once::new();
...
+            // .without_time()
```
Comment (leonardoalt) [inline r55/src/test_utils.rs:19]: Leftover?

File: r55/src/exec.rs [hunk 5]
```diff
@@ -133,32 +136,89 @@ pub fn handle_register<EXT, DB: Database>(handler: &mut EvmHandler<'_, EXT, DB>)
...
 fn execute_riscv(
     rvemu: &mut RVEmu,
     interpreter: &mut Interpreter,
-    shared_memory: &mut SharedMemory,
+    _shared_memory: &mut SharedMemory,
     host: &mut dyn Host,
 ) -> Result<InterpreterAction> {
+    debug!(
+        "{} RISC-V execution:  PC: {:#x}  Return data dst: {:#?}",
+        if rvemu.emu.cpu.pc == 0x80300000 {
```
Comment (leonardoalt) [inline r55/src/exec.rs:206]: This comes from the linker script, right? It'd be better as a constant

File: r55/src/exec.rs [hunk 6]
```diff
@@ -250,45 +310,89 @@ fn execute_riscv(
...
                                 caller: interpreter.contract.target_address,
-                                value: CallValue::Transfer(U256::from_le_bytes(
-                                    value.to_le_bytes(),
-                                )),
+                                value: CallValue::Transfer(U256::from(value)),
                                 scheme: CallScheme::Call,
                                 is_static: false,
                                 is_eof: false,
-                                return_memory_offset: 0..ret_size as usize,
+                                return_memory_offset: 0..0, // handled with `returned_data_destiny`
```
Comment (leonardoalt) [inline r55/src/exec.rs:395]: I wonder if we need this when a RISCV contract calls an EVM contract

File: r55/src/exec.rs [hunk 7]
```diff
@@ -442,9 +546,14 @@ fn execute_riscv(
                     }
                 }
             }
-            _ => {
-                let total_cost = r55_gas_used(&emu.cpu.inst_counter) + evm_gas;
-                return return_revert(interpreter, total_cost);
+            Ok(_) => {
+                debug!("Successful instruction at PC: {:#x}", emu.cpu.pc);
```
Comment (leonardoalt) [inline r55/src/exec.rs:550]: Maybe this should be trace instead of debug?

File: r55/src/exec.rs [hunk 8]
```diff
@@ -133,32 +136,89 @@ pub fn handle_register<EXT, DB: Database>(handler: &mut EvmHandler<'_, EXT, DB>)
...
+                                .cpu
+                                .bus
+                                .get_dram_slice(return_range.clone())
+                                .expect("unable to get memory from return range");
+
+                            debug!("- Return data: {:?}", res.output);
+                            debug!("- Memory range: {:?}", return_range);
+                            debug!("- Memory size: {}", return_memory.len());
+
+                            // write return data to parent's memory
```
Comment (leonardoalt) [inline r55/src/exec.rs:181]: This only supports the case where a RISCV contract calls a RISCV contract, right?

File: contract-derive/src/lib.rs [hunk 9]
```diff
@@ -275,4 +274,126 @@ fn is_payable(method: &syn::ImplItemMethod) -> bool {
         }
         false
     })
-}
\ No newline at end of file
+}
+
+#[proc_macro_attribute]
+pub fn interface(_attr: TokenStream, item: TokenStream) -> TokenStream {
```
Comment (leonardoalt) [inline contract-derive/src/lib.rs:280]: Sick!

**Review by leonardoalt — COMMENTED** [review]
This is awesome!

Left a few small comments.
Also can you remove the `rvemu.*` files (we should maybe add them to gitignore).

File: r55/src/test_utils.rs (hunk 3 above)
Comment (0xrusowsky) [reply r55/src/test_utils.rs:18]: this specific line allows the user to define the log level (`debug`, `info`, `warn`, or `error`) via the env var `RUST_LOG`

File: r55/src/exec.rs (hunk 5 above)
Comment (0xrusowsky) [reply r55/src/exec.rs:206]: yes, will do

File: contract-derive/src/lib.rs (hunk 9 above)
Comment (0xrusowsky) [reply contract-derive/src/lib.rs:280]: heavy inspired by your contract macro haha

File: r55/src/exec.rs (hunk 8 above)
Comment (0xrusowsky) [reply r55/src/exec.rs:181]: yeah, i totally missed that

File: r55/src/exec.rs (hunk 1 above)
Comment (0xrusowsky) [reply r55/src/exec.rs:188]: based on @leonardoalt comments it looks like my implementation only works with RISCV contracts, as we do need to use shared memory when interacting with EVM contracts

**Review by leonardoalt — APPROVED** [review]

**Review by leonardoalt — APPROVED** [review]
